Compilation fails with spack with cce compilers

Hello,

I am trying to compile open openblas on a cluster with the cce compilers, but it fails:
This is the spack info:
```
> spack -V
0.20.0
```
```
> spack compilers
==> Available compilers
-- cce sles15-any -----------------------------------------------
cce@15.0.1  cce@14.0.2

-- clang sles15-any ---------------------------------------------
clang@14.0.0-rocm5.2.3

-- gcc sles15-any -----------------------------------------------
gcc@7.5.0  gcc@12.2.0  gcc@11.2.0
``` 
I try to install it via
``` 
spack install openblas%cce@15.0.1
```  
The build fails with this message:
```
Installing openblas-0.3.23-3p5f6xweur4fli4hx4oqlwngfanjhnic
....
....
.....
 Error: ProcessError: Command exited with status 2:
    'make' '-j32' '-s' 'CC=/pfs/lustrep1/appl/lumi/spack/23.03/0.20.0-user/lib/spack/env/cce/craycc' 'FC=/pfs/lustrep1/appl/lumi/spack/23.03/0.20.0-user/lib/spack/env/cce/crayftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'all'

4 errors found in build log:
     11    1 warning generated.
     12    ar: creating ../libopenblas_zen-r0.3.23.a
     13    parameter.c:273:7: warning: unused variable 'size' [-Wunused-variable]
     14      int size = 16;
     15          ^
     16    1 warning generated.
  >> 17    ld.lld: error: undefined reference due to --no-allow-shlib-undefined: la_constants_
     18    >>> referenced by ../libopenblas_zen-r0.3.23.so
  >> 19    clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 20    make[1]: *** [Makefile:196: ../libopenblas_zen-r0.3.23.so] Error 1
  >> 21    make: *** [Makefile:132: shared] Error 2
``` 

The complete log is this:
``` 
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> [2023-09-14-08:47:31.908689] 'make' '-j32' '-s' 'CC=/pfs/lustrep1/appl/lumi/spack/23.03/0.20.0-user/lib/spack/env/cce/craycc' 'FC=/pfs/lustrep1/appl/lumi/spack/23.03/0.20.0-user/lib/spack/env/cce/crayftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'all'
gemmt.c:99:9: warning: unused variable 'alpha' [-Wunused-variable]
        FLOAT *alpha = Alpha;
               ^
1 warning generated.
gemmt.c:99:9: warning: unused variable 'alpha' [-Wunused-variable]
        FLOAT *alpha = Alpha;
               ^
1 warning generated.
ar: creating ../libopenblas_zen-r0.3.23.a
parameter.c:273:7: warning: unused variable 'size' [-Wunused-variable]
  int size = 16;
      ^
1 warning generated.
ld.lld: error: undefined reference due to --no-allow-shlib-undefined: la_constants_
>>> referenced by ../libopenblas_zen-r0.3.23.so
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [Makefile:196: ../libopenblas_zen-r0.3.23.so] Error 1
make: *** [Makefile:132: shared] Error 2
```

I just tried this 
```
make PREFIX=/scratch/project_462000007/cristian/BLASSOURCE/ LIBNAMESUFFIX=nonthreaded   -j DYNAMIC_ARCH=0 CC=cc FC=ftn HOSTCC=gcc BINARY=64 INTERFACE=64 NO_AFFINITY=1 NO_WARMUP=1 USE_OPENMP=0 USE_THREAD=0 USE_LOCKING=1 LIBNAMESUFFIX=nonthreaded
``` 
I get this error:
```
ld.lld: error: undefined reference due to --no-allow-shlib-undefined: la_constants_
>>> referenced by ../libopenblas_nonthreaded_zen-r0.3.24.so
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [Makefile:196: ../libopenblas_nonthreaded_zen-r0.3.24.so] Error 1
```

It seems that it compiled. Though some tests failed. Should I be worried about that.
```
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c zblat3.f  -o zblat3.o
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
make[1]: *** [../Makefile.system:1813: sblat1.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [../Makefile.system:1813: zblat1.o] Error 1
make[1]: *** [../Makefile.system:1813: dblat2.o] Error 1
make[1]: *** [../Makefile.system:1813: dblat1.o] Error 1
make[1]: *** [../Makefile.system:1813: zblat2.o] Error 1
make[1]: *** [../Makefile.system:1813: cblat1.o] Error 1
make[1]: *** [../Makefile.system:1813: sblat2.o] Error 1
make[1]: *** [../Makefile.system:1813: cblat2.o] Error 1
make[1]: *** [../Makefile.system:1813: sblat3.o] Error 1
make[1]: *** [../Makefile.system:1813: dblat3.o] Error 1
make[1]: *** [../Makefile.system:1813: zblat3.o] Error 1
make[1]: *** [../Makefile.system:1813: cblat3.o] Error 1
make[1]: Leaving directory '/pfs/lustrep4/scratch/project_462000007/cristian/BLASSPACK/OpenBLAS/test'
make: *** [Makefile:156: tests] Error 2
```

I tried your suggestion with this compilation:

```
make clean;
 make PREFIX=/scratch/project_462000007/cristian/BLASSOURCE/ LIBNAMESUFFIX=nonthreaded   -j DYNAMIC_ARCH=0 CC=cc FC=ftn HOSTCC=gcc BINARY=64 INTERFACE=64 NO_AFFINITY=1 NO_WARMUP=1 USE_OPENMP=0 USE_THREAD=0 USE_LOCKING=1 LIBNAMESUFFIX=nonthreaded
```

I still getting the same errors:
```
make[1]: Entering directory '/pfs/lustrep4/scratch/project_462000007/cristian/BLASSPACK/OpenBLAS/test'
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c sblat1.f  -o sblat1.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c dblat1.f  -o dblat1.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c cblat1.f  -o cblat1.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c zblat1.f  -o zblat1.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c sblat2.f  -o sblat2.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c dblat2.f  -o dblat2.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c cblat2.f  -o cblat2.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c zblat2.f  -o zblat2.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c sblat3.f  -o sblat3.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c dblat3.f  -o dblat3.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c cblat3.f  -o cblat3.o
ftn -O2 -hnopattern -O noomp -fPIC -msse3 -mssse3 -msse4.1 -mavx -mavx2 -c zblat3.f  -o zblat3.o
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
ftn-2307 ftn: ERROR in command line
  The "-m" option must be followed by 0, 1, 2, 3 or 4.
make[1]: *** [../Makefile.system:1813: sblat1.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [../Makefile.system:1813: dblat1.o] Error 1
make[1]: *** [../Makefile.system:1813: cblat1.o] Error 1
make[1]: *** [../Makefile.system:1813: sblat2.o] Error 1
make[1]: *** [../Makefile.system:1813: zblat1.o] Error 1
make[1]: *** [../Makefile.system:1813: cblat2.o] Error 1
make[1]: *** [../Makefile.system:1813: dblat2.o] Error 1
make[1]: *** [../Makefile.system:1813: zblat2.o] Error 1
make[1]: *** [../Makefile.system:1813: dblat3.o] Error 1
make[1]: *** [../Makefile.system:1813: cblat3.o] Error 1
make[1]: *** [../Makefile.system:1813: sblat3.o] Error 1
make[1]: *** [../Makefile.system:1813: zblat3.o] Error 1
make[1]: Leaving directory '/pfs/lustrep4/scratch/project_462000007/cristian/BLASSPACK/OpenBLAS/test'
make: *** [Makefile:156: tests] Error 2
``` 
More info:

```
> module list

Currently Loaded Modules:
  1) craype-x86-rome      4) perftools-base/22.12.0                  7) craype/2.7.19      10) cray-libsci/22.12.1.1      13) lumi-tools/23.04 (S)
  2) libfabric/1.15.2.0   5) xpmem/2.5.2-2.4_3.47__gd0f7936.shasta   8) cray-dsmml/0.2.2   11) PrgEnv-cray/8.3.3          14) init-lumi/0.2    (S)
  3) craype-network-ofi   6) cce/15.0.0                              9) cray-mpich/8.1.23  12) ModuleLabel/label     (S)
```

--- a/Makefile
+++ b/Makefile
@@ -35,7 +35,11 @@ export NO_LAPACK
 export C_LAPACK
 endif
 
+ifeq ($(F_COMPILER),CRAY)
+LAPACK_NOOPT := $(filter-out -O0 -O1 -O2 -O3 -Ofast -Og -Os,$(LAPACK_FFLAGS))
+else
 LAPACK_NOOPT := $(filter-out -O0 -O1 -O2 -O3 -Ofast -O -Og -Os,$(LAPACK_FFLAGS))
+endif
 
 SUBDIRS_ALL = $(SUBDIRS) test ctest utest exports benchmark ../laswp ../bench cpp_thread_test
 

--- a/Makefile.system
+++ b/Makefile.system
@@ -1642,11 +1642,11 @@ endif
 
 ifeq ($(F_COMPILER),NAG)
 LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
-FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
+override FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
 endif
 ifeq ($(F_COMPILER),CRAY)
 LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
-FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
+override FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
 endif
 
 LAPACK_CFLAGS = $(CFLAGS)

